File: scala/allergies/Allergies.scala
import Allergen.Allergen

class Allergies {
  private lazy val allergens = Allergen.values.toList

  def isAllergicTo(allergen: Allergen, score: Int) = (allergen.id & score) != 0
  def allergies(score: Int) = allergens.filter(isAllergicTo(_, score))
}

object Allergies {
  def apply() = new Allergies
}

object Allergen extends Enumeration {
  type Allergen = Value
  val Eggs = Value(1 << 0)
  val Peanuts = Value(1 << 1)
  val Shellfish = Value(1 << 2)
  val Strawberries = Value(1 << 3)
  val Tomatoes = Value(1 << 4)
  val Chocolate = Value(1 << 5)
  val Pollen = Value(1 << 6)
  val Cats = Value(1 << 7)
}

File: nim/allergies/allergies.nim
import bitops, tables, sequtils

const Allergens = 
  {
    "eggs":         0,
    "peanuts":      1,
    "shellfish":    2,
    "strawberries": 3,
    "tomatoes":     4,
    "chocolate":    5,
    "pollen":       6,
    "cats":         7
  }.toOrderedTable

type
  Allergies* = object
    score*: int

proc isAllergicTo*(allergies: Allergies, allergen: string): bool =
  testBit(allergies.score, Allergens[allergen])

proc lst*(allergies: Allergies): seq[string] =
  toSeq(Allergens.keys).filterIt(allergies.isAllergicTo(it))
